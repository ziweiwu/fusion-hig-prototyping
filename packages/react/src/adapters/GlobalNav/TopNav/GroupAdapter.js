import React from "react";
import PropTypes from "prop-types";
import { GlobalNav as VanillaGlobalNav } from "hig-vanilla";
import HIGAdapter, {
  MountedByHIGParentList,
  MountsHIGChildList
} from "../../HIGAdapter";

function GroupAdapter(props) {
  return (
    <HIGAdapter
      {...props}
      displayName="Group"
      HIGConstructor={
        VanillaGlobalNav._partials.TopNav._partials.Help._partials.Group
      }
    >
      {adapterProps => (
        <div>
          <MountedByHIGParentList mounter="addGroup" {...adapterProps}>
            {props.children}
          </MountedByHIGParentList>
          <MountsHIGChildList {...adapterProps}>
            {props.children}
          </MountsHIGChildList>
        </div>
      )}
    </HIGAdapter>
  );
}

GroupAdapter.propTypes = {
  children: PropTypes.node
};

GroupAdapter.defaultProps = {
  children: undefined
};

export default GroupAdapter;
