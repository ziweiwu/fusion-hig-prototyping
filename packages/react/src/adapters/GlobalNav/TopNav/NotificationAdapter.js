import React, { Component } from "react";
import PropTypes from "prop-types";
import { Notification as VanillaNotification } from "hig-vanilla";

import HIGAdapter, {
  MountedByHIGParent,
  MapsPropToMethod,
  MountsAnyChild,
  MapsEventListener
} from "../../HIGAdapter";

export default class NotificationAdapter extends Component {
  render() {
    return (
      <HIGAdapter
        displayName="NotificationAdapter"
        HIGConstructor={VanillaNotification}
        {...this.props}
      >
        {adapterProps => (
          <div>
            <MountedByHIGParent mounter="addNotification" {...adapterProps} />
            <MapsPropToMethod
              value={this.props.timestamp}
              setter="setCreatedAt"
              {...adapterProps}
            />
            <MapsEventListener
              listener="onLinkClick"
              handler={this.props.onLinkClick}
              {...adapterProps}
            />
            <MapsEventListener
              listener="onDismiss"
              handler={this.props.onDismiss}
              {...adapterProps}
            />
            <MapsPropToMethod value={this.props.unread} {...adapterProps}>
              {(instance, value) =>
                value ? instance.markUnread() : instance.markRead()
              }
            </MapsPropToMethod>
            <MapsPropToMethod value={this.props.featured} {...adapterProps}>
              {(instance, value) =>
                value ? instance.setFeatured() : instance.removeFeatured()
              }
            </MapsPropToMethod>
            <MountsAnyChild mounter="setContent" {...adapterProps}>
              {this.props.children}
            </MountsAnyChild>
          </div>
        )}
      </HIGAdapter>
    );
  }
}

NotificationAdapter.propTypes = {
  /**
   * A callback called when user clicks on a link in the notification
   */
  onLinkClick: PropTypes.func,
  /**
   * A callback called when user dismisses a featured notification
   */
  onDismiss: PropTypes.func,
  /**
   * {Boolean} to show specify whether notificaiton is read
   */
  unread: PropTypes.bool,
  /**
   * Indicates whether this is a featured notification
   */
  featured: PropTypes.bool,
  /**
   * Content for notification
   */
  children: PropTypes.node,
  /**
   * Timestamp for notification
   */
  timestamp: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Date)]) // ISO date string
};
