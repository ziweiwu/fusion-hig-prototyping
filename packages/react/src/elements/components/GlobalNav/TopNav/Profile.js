import React, { Component } from "react";
import ProfileAdapter from "../../../../adapters/GlobalNav/TopNav/ProfileAdapter";

class Profile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false
    };
  }

  handleProfileImageClick = () => {
    this.setState({ open: !this.state.open });
  };

  handleProfileClickOutside = () => {
    this.setState({ open: false });
  };

  render() {
    return (
      <ProfileAdapter
        open={this.state.open}
        {...this.props}
        onProfileImageClick={this.handleProfileImageClick}
        onProfileClickOutside={this.handleProfileClickOutside}
      />
    );
  }
}

export default Profile;
