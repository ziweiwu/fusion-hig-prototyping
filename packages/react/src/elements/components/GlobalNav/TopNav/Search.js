import React, { Component } from "react";
import PropTypes from "prop-types";
import SearchAdapter from "../../../../adapters/GlobalNav/TopNav/SearchAdapter";
import Option from "../../FormElements/Option";

export default class Search extends Component {
  static defaultProps = {
    onBlur: () => {},
    onClearIconClick: () => {},
    onClickOutside: () => {},
    onFocus: () => {},
    onInput: () => {},
    onKeydown: () => {},
    onOptionSelect: () => {},
    onSubmit: () => {},
    showClearIcon: undefined,
    showOptions: undefined,
    value: undefined
  };

  constructor(props) {
    super(props);

    const controlled = props.value !== undefined;

    this.state = {
      showOptions: false,
      value: undefined,
      controlled
    };
  }

  onInput = event => {
    this.props.onInput({ value: event.target.value });
    const clearIconVisible = event.target.value.length > 0;
    this.setState({ showOptions: true, clearIconVisible });
  };

  setValue(value) {
    if (this.state.controlled) {
      this.setState({ value: this.state.value });
    } else {
      this.setState({ value });
    }
  }

  handleOptionSelect = () => {
    const selectedOption = this.props.options[this.state.focusedOptionIndex];

    this.props.onOptionSelect(selectedOption);
    this.hideOptions();
  };

  handleOptionHover = focusedOptionIndex => () =>
    this.setState({ focusedOptionIndex });

  hideOptions = () => {
    this.setState({ showOptions: false, focusedOptionIndex: undefined });
  };

  showOptions = () => {
    const value =
      this.props.options !== undefined
        ? this.props.options.length > 0 && this.state.showOptions
        : false;
    return value;
  };

  handleClearIconClick = () => {
    this.setValue("");
    this.props.onInput({ value: "" });
    this.setState({ focusedOptionIndex: undefined, clearIconVisible: false });
  };

  handleKeydown = event => {
    switch (event.keyCode) {
      case 40:
        event.preventDefault();
        this.focusPrevious();
        break;
      case 37:
        event.preventDefault();
        this.focusPrevious();
        break;
      case 38:
        event.preventDefault();
        this.focusNext();
        break;
      case 39:
        event.preventDefault();
        this.focusNext();
        break;
      case 13:
        this.handleEnterPress(event.target.value);
        break;
      default:
        break;
    }
  };

  handleEnterPress(selectedValue) {
    if (this.state.focusedOptionIndex !== undefined) {
      this.handleOptionSelect(selectedValue);
      return;
    }

    this.props.onSubmit({ selectedValue });
    this.hideOptions();
  }

  focusPrevious() {
    if (
      this.state.focusedOptionIndex === undefined ||
      this.state.focusedOptionIndex >= this.props.options.length - 1
    ) {
      this.setState({ focusedOptionIndex: 0, showOptions: true });
    } else {
      const focusedOptionIndex = this.state.focusedOptionIndex + 1;
      this.setState({ focusedOptionIndex, showOptions: true });
    }
  }

  focusNext() {
    if (
      this.state.focusedOptionIndex === undefined ||
      this.state.focusedOptionIndex <= 0
    ) {
      const focusedOptionIndex = this.props.options.length - 1;
      this.setState({ focusedOptionIndex, showOptions: true });
    } else {
      const focusedOptionIndex = this.state.focusedOptionIndex - 1;
      this.setState({ focusedOptionIndex, showOptions: true });
    }
  }

  render() {
    return (
      <SearchAdapter
        {...this.props}
        value={this.props.value}
        onInput={this.onInput}
        showOptions={this.showOptions()}
        showClearIcon={this.state.clearIconVisible}
        onClearIconClick={this.handleClearIconClick}
        onClickOutside={this.hideOptions}
        onKeydown={this.handleKeydown}
      >
        {this.props.options !== undefined && this.props.options.length > 0
          ? this.props.options.map((option, index) => (
              <Option
                key={option.value}
                focused={index === this.state.focusedOptionIndex}
                {...option}
                onClick={this.handleOptionSelect}
                onHover={this.handleOptionHover(index)}
              />
            ))
          : null}
      </SearchAdapter>
    );
  }
}

Search.propTypes = {
  /**
   * Autocomplete options to present when begins typing their query
   */
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.node,
      value: PropTypes.string
    })
  ),
  /**
   * Calls the provided callback when user enters text in Search input
   */
  onInput: PropTypes.func,
  /**
   * Value of search input
   */
  value: PropTypes.string,
  /**
   * Shows filtered options
   */
  showOptions: PropTypes.bool,
  /**
   * Calls the callback on Blur
   */
  onBlur: PropTypes.func,
  /**
   * Calls the callback on Focus
   */
  onFocus: PropTypes.func,
  /**
   * Show the clear input icon
   */
  showClearIcon: PropTypes.bool,
  /**
   * Calls the provided callback when the clear icon is clicked
   */
  onClearIconClick: PropTypes.func,
  /**
   * Calls the provided callback when user clicks outside of the auto-complete menu
   */
  onClickOutside: PropTypes.func,
  /**
   * Calls the provided callback when user presses a key when the Search has focus
   */
  onKeydown: PropTypes.func,
  /**
   * Calls the provided callback when user presses Enter key while the Search has focus
   */
  onSubmit: PropTypes.func,
  /**
   * Calls the provided callback when user selects an auto-complete option
   */
  onOptionSelect: PropTypes.func
};
