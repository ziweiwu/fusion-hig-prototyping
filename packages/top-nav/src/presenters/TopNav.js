import React from "react";
import PropTypes from "prop-types";

export default function TopNav({ leftActions, rightActions, logo }) {
  return (
    <div className="hig__top-nav">
      {leftActions}
      <div className="hig__top-nav__item">
        <div className="hig__top-nav__logo-wrapper">{logo}</div>
      </div>
      <div className="hig__top-nav__spacer" />
      {rightActions}
    </div>
  );
}

TopNav.propTypes = {
  /** Actions to render to the left of the logo */
  leftActions: PropTypes.node,
  /** Actions to render on the right of the bar */
  rightActions: PropTypes.node,
  /** Content to render as the logo */
  logo: PropTypes.node
};
