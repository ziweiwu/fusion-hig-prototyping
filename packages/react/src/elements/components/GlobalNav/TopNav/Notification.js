import React, { Component } from "react";
import PropTypes from "prop-types";

import NotificationAdapter from "../../../../adapters/GlobalNav/TopNav/NotificationAdapter";

export default class Notification extends Component {
  constructor(props) {
    super(props);
    this.state = {
      unread: this.props.unread
    };
    this.initialProps = props;
  }

  onLinkClick = () => {
    this.props.onLinkClick(this.props.id);
  };

  render() {
    return (
      <NotificationAdapter
        {...this.props}
        onLinkClick={this.onLinkClick}
        id={this.props.id}
        unread={this.props.unread}
      >
        {this.props.children}
      </NotificationAdapter>
    );
  }
}

export function notificationElementPropType(props, propName) {
  const value = props[propName];
  if (value && value.type && value.type === Notification.name) {
    return new Error("Expected an element of type Notification.");
  }
  return null;
}

Notification.defaultProps = {
  onLinkClick: () => {}
};

Notification.propTypes = {
  /**
   * {Boolean} to show specify whether notification is read
   */
  unread: PropTypes.bool,
  /**
   * Content for notification
   */
  children: PropTypes.node,
  /**
   * Timestamp for notification (ISO date string or date instance)
   */
  timestamp: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.instanceOf(Date)
  ]),
  /**
   * A callback called when user clicks on a link in the notification
   */
  onLinkClick: PropTypes.func,
  /**
   * A callback called when user dismisses a featured notification
   */
  onDismiss: PropTypes.func,
  /**
   * Unique Id for the notification
   */
  id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  /**
   * {Boolean} specifies a featured notification
   */
  featured: PropTypes.bool
};
