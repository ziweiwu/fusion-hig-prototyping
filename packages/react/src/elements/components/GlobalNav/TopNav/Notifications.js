import React, { Component } from "react";
import PropTypes from "prop-types";

import NotificationsAdapter from "../../../../adapters/GlobalNav/TopNav/NotificationsAdapter";
import { notificationElementPropType } from "./Notification";

export default class Notifications extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: this.props.open,
      seenNotificationIds: this.readIds()
    };
  }

  onScroll = event => this.props.onScroll && this.props.onScroll(event);

  onClick = event => {
    if (this.props.onClick) {
      this.props.onClick({
        event
      });
    }

    this.setState({
      open: true
    });
  };

  onClickOutside = event => {
    this.setState({
      open: false,
      seenNotificationIds: this.notificationIds(),
      featuredNotificationSeen: true
    });

    if (this.props.onClickOutside) {
      this.props.onClickOutside(event);
    }
  };

  readIds() {
    return React.Children.toArray(this.props.children)
      .filter(c => !c.props.unread)
      .map(c => c.props.id);
  }

  unseenCount = () => {
    let count =
      this.props.featuredNotification && !this.state.featuredNotificationSeen
        ? 1
        : 0;

    if (!this.props.children) {
      return count;
    }

    count += this.notificationIds().filter(
      id => this.state.seenNotificationIds.indexOf(id) === -1
    ).length;

    return Math.min(count, 99);
  };

  notificationIds() {
    return React.Children.map(this.props.children, child => child.props.id);
  }

  render() {
    const unseenCount = this.unseenCount();
    return (
      <NotificationsAdapter
        {...this.props}
        onClick={this.onClick}
        onClickOutside={this.onClickOutside}
        onScroll={this.onScroll}
        open={this.state.open}
        unseenCount={unseenCount}
        showNotificationsCount={unseenCount > 0}
      >
        {this.props.featuredNotification &&
          React.cloneElement(this.props.featuredNotification, {
            featured: true,
            unread: false // Hide unread indicator on featured notifications
          })}
        {this.props.children}
      </NotificationsAdapter>
    );
  }
}

Notifications.propTypes = {
  /**
   * Opens the notifications flyout
   */
  open: PropTypes.bool,
  /**
   * Show the loading indicator
   */
  loading: PropTypes.bool,
  /**
   * Calls the provided callback when user clicks on the notifications icon in the top nav
   */
  onClick: PropTypes.func,
  /**
   * Calls the provided callback when user clicks outside the dropdown
   */
  onClickOutside: PropTypes.func,
  /**
   * Calls the provided callback when the notifications content is scrolled
   */
  onScroll: PropTypes.func,
  /**
   * The title text that renders above the notifications list
   */
  title: PropTypes.string,
  /**
   * A Notification component, to be styled as a featured notification
   */
  featuredNotification: notificationElementPropType
};

Notifications.defaultProps = {
  onClick: () => {},
  onClickOutside: () => {},
  onScroll: () => {},
  title: "Notifications",
  children: []
};
