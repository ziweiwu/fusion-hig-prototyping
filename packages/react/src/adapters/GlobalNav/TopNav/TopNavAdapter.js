import React from "react";
import PropTypes from "prop-types";
import { GlobalNav as VanillaGlobalNav } from "hig-vanilla";
import HIGAdapter, {
  MapsPropToMethod,
  MapsEventListener,
  MountedByHIGParent,
  MountsHIGChild
} from "../../HIGAdapter";
import ShortcutAdapter from "./ShortcutAdapter";
import Search from "../../../elements/components/GlobalNav/TopNav/Search";
import Notifications from "../../../elements/components/GlobalNav/TopNav/Notifications";
import Profile from "../../../elements/components/GlobalNav/TopNav/Profile";
import HelpAdapter from "./HelpAdapter";
import ProjectAccountSwitcherAdapter from "./ProjectAccountSwitcherAdapter";
import ProjectAccountSwitcher from "../../../elements/components/GlobalNav/TopNav/ProjectAccountSwitcher"; // eslint-disable-line max-len

function sortChildren(children) {
  return {
    notifications: children.find(child => child.type === Notifications),
    shortcut: children.find(child => child.type === ShortcutAdapter),
    search: children.find(child => child.type === Search),
    profile: children.find(child => child.type === Profile),
    help: children.find(child => child.type === HelpAdapter),
    projectAccountSwitcher: children.find(child =>
      [ProjectAccountSwitcher, ProjectAccountSwitcherAdapter].includes(
        child.type
      )
    )
  };
}

function TopNavAdapter(props) {
  const {
    notifications,
    shortcut,
    search,
    profile,
    help,
    projectAccountSwitcher
  } = sortChildren(React.Children.toArray(props.children));

  return (
    <HIGAdapter
      displayName="TopNav"
      HIGConstructor={VanillaGlobalNav._partials.TopNav}
      {...props}
    >
      {adapterProps => (
        <div>
          <MountedByHIGParent mounter="addTopNav" {...adapterProps} />
          <MapsPropToMethod
            value={props.logo}
            setter="setLogo"
            {...adapterProps}
          />
          <MapsPropToMethod
            value={props.logoLink}
            setter="setLogoLink"
            {...adapterProps}
          />
          <MapsPropToMethod
            value={props.hideHamburgerButton}
            setter="hideMenu"
            {...adapterProps}
          />
          <MapsEventListener
            handler={props.onLogoClick}
            listener="onLogoClick"
            {...adapterProps}
          />
          <MapsEventListener
            handler={props.onHamburgerClick}
            listener="onHamburgerClick"
            {...adapterProps}
          />
          <MountsHIGChild {...adapterProps}>{shortcut}</MountsHIGChild>
          <MountsHIGChild {...adapterProps}>{search}</MountsHIGChild>
          <MountsHIGChild {...adapterProps}>{profile}</MountsHIGChild>
          <MountsHIGChild {...adapterProps}>{notifications}</MountsHIGChild>
          <MountsHIGChild {...adapterProps}>{help}</MountsHIGChild>
          <MountsHIGChild {...adapterProps}>
            {projectAccountSwitcher}
          </MountsHIGChild>
        </div>
      )}
    </HIGAdapter>
  );
}

TopNavAdapter.propTypes = {
  children: PropTypes.node,
  hideHamburgerButton: PropTypes.bool,
  logo: PropTypes.string,
  logoLink: PropTypes.string,
  onLogoClick: PropTypes.func,
  onHamburgerClick: PropTypes.func
};

TopNavAdapter.defaultProps = {
  children: undefined,
  hideHamburgerButton: false,
  logo: undefined,
  logoLink: undefined,
  onLogoClick: undefined,
  onHamburgerClick: undefined
};

export default TopNavAdapter;
