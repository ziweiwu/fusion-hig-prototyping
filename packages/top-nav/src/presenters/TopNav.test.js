import React from "react";
import renderer from "react-test-renderer";

import TopNav from "./TopNav";

describe("top-nav/TopNav", () => {
  const cases = [
    {
      description: "renders with no props",
      props: {}
    },
    {
      description: "renders with actions and logo",
      props: {
        leftActions: <div className="leftActions" />,
        rightActions: <div className="rightActions" />,
        logo: <div className="logo" />
      }
    }
  ];

  cases.forEach(({ description, props }) => {
    it(description, () => {
      const tree = renderer.create(<TopNav {...props} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
