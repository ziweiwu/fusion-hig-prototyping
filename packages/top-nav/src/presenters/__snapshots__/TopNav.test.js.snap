// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`top-nav/TopNav renders with actions and logo 1`] = `
<div
  className="hig__top-nav"
>
  <div
    className="leftActions"
  />
  <div
    className="hig__top-nav__item"
  >
    <div
      className="hig__top-nav__logo-wrapper"
    >
      <div
        className="logo"
      />
    </div>
  </div>
  <div
    className="hig__top-nav__spacer"
  />
  <div
    className="rightActions"
  />
</div>
`;

exports[`top-nav/TopNav renders with no props 1`] = `
<div
  className="hig__top-nav"
>
  <div
    className="hig__top-nav__item"
  >
    <div
      className="hig__top-nav__logo-wrapper"
    />
  </div>
  <div
    className="hig__top-nav__spacer"
  />
</div>
`;
